server/prometheus.js

```diff
@@ -1,63 +1,125 @@
 const PrometheusClient = require("prom-client");
 const { log } = require("../src/util");
+const { R } = require("redbean-node");
 
-const commonLabels = [
-    "monitor_id",
-    "monitor_name",
-    "monitor_type",
-    "monitor_url",
-    "monitor_hostname",
-    "monitor_port",
-];
-
-const monitorCertDaysRemaining = new PrometheusClient.Gauge({
-    name: "monitor_cert_days_remaining",
-    help: "The number of days remaining until the certificate expires",
-    labelNames: commonLabels
-});
-
-const monitorCertIsValid = new PrometheusClient.Gauge({
-    name: "monitor_cert_is_valid",
-    help: "Is the certificate still valid? (1 = Yes, 0= No)",
-    labelNames: commonLabels
-});
-const monitorResponseTime = new PrometheusClient.Gauge({
-    name: "monitor_response_time",
-    help: "Monitor Response Time (ms)",
-    labelNames: commonLabels
-});
-
-const monitorStatus = new PrometheusClient.Gauge({
-    name: "monitor_status",
-    help: "Monitor Status (1 = UP, 0= DOWN, 2= PENDING, 3= MAINTENANCE)",
-    labelNames: commonLabels
-});
+let monitorCertDaysRemaining = null;
+let monitorCertIsValid = null;
+let monitorResponseTime = null;
+let monitorStatus = null;
 
 class Prometheus {
     monitorLabelValues = {};
 
     /**
      * @param {object} monitor Monitor object to monitor
+     * @param {Array<{name:string,value:?string}>} tags Tags to add to the monitor
      */
-    constructor(monitor) {
+    constructor(monitor, tags) {
         this.monitorLabelValues = {
             monitor_id: monitor.id,
             monitor_name: monitor.name,
             monitor_type: monitor.type,
             monitor_url: monitor.url,
             monitor_hostname: monitor.hostname,
-            monitor_port: monitor.port
+            monitor_port: monitor.port,
+            ...this.mapTagsToLabels(tags)
         };
     }
 
+    /**
+     * Initialize Prometheus metrics, and add all available tags as possible labels.
+     * This should be called once at the start of the application.
+     * New tags will NOT be added dynamically, a restart is sadly required to add new tags to the metrics.
+     * Existing tags added to monitors will be updated automatically.
+     * @returns {Promise<void>}
+     */
+    static async init() {
+        const tags = (await R.findAll("tag")).map((tag) => {
+            return Prometheus.sanitizeForPrometheus(tag.name);
+        }).filter((tagName) => {
+            return tagName !== "";
+        });
```

Comment:
🤔 this does not impose a distinct constraint on the tag names => If I have `foo=bar` and `foo=baz`, foo is now twice in this list.

Will need to test if this causes problems in the prometheus code.